sympy/solvers/solveset.py

```diff
@@ -978,6 +985,342 @@ def _solveset(f, symbol, domain, _check=False):
+    Helper to check whether an equation is exponential or not.
+
+    ``f`` is the equation to be checked
+
+    ``symbol`` is the variable in which the equation needs to be checked
+
+    The function extracts each term of the equation and checks if it is
+    of exponential form w.r.t `symbol`.
+    Returns `True` if any term is of exponential type otherwise `False`.
+
+    Examples
+    ========
+
+    >>> from sympy import symbols, cos, exp
+    >>> from sympy.solvers.solveset import _check_expo as check
+    >>> x, y = symbols('x y')
+    >>> check(x**y - x, y)
+    True
+    >>> check(x**y - x, x)
+    False
+    >>> check(exp(x + 3), x)
+    True
+    >>> check(cos(2**x), x)
+    False
+    """
+
+    for add_arg in Add.make_args(f):
+        for mul_arg in Mul.make_args(add_arg):
+            if isinstance(mul_arg, (Pow, exp)) and \
+                    symbol in mul_arg.exp.free_symbols:
+                return True
+    return False
+
+
+def _transolve(f, symbol, domain):
+    r"""
+    Function to solve transcendental equations. It is a helper to
+    `solveset` and should be used internally as of now. `\_transolve`
+    currently supports the following class of equations:
+
+        - Exponential equations
+
+    Parameters
+    ==========
+
+    f : can be any transcendental equation that needs to be solved.
+        This needs to be an expression, which is assumed
+        to be equal to 0.
+
+    symbol : is variable for which the equation is solved.
+        This needs to be of class `Symbol`.
+
+    domain : is a set over which the equation is solved.
+        This needs to be of class `Set`.
+
+    Examples
+    ========
+
+    >>> from sympy.solvers.solveset import _transolve as transolve
+    >>> from sympy.solvers.solvers import _tsolve as tsolve
+    >>> from sympy import symbols, S, pprint
+    >>> x = symbols('x', real=True)
+    >>> transolve(5**(x - 3) - 3**(2*x + 1), x, S.Reals)
+    {-log(375)/(-log(5) + 2*log(3))}
+
+    How to use `\_transolve`
+    ========================
+
+    `\_transolve` should be used as an internal function. There should be
+    an internal call to `\_transolve`, passing all the three above
+    parameters as arguments. These three arguments should be sympy
+    instances.
+
+    How `\_transolve` works
+    =======================
+
+    `\_transolve` uses two types of helper functions to solve equation
+    of a particular class:
+
+    Identifying helpers: These helpers are used to determine whether a
+    given equation belongs to a certain class of equation or not.
+    These functions return either True or False.
+
+    Solving helpers: These are the helpers that solve the equation. They
+    are invoked once the equation is identified by the `identifying
+    helpers`. These function return either the complete solution or
+    modifies the equation for `solveset` to handle.
+
+    * Philosophy behind the module
+
+    `\_transolve` comes into action when `solveset` is unable to solve
```

Comment:
```
>>> transolve((2**x - 4**x)/(3**x),x,S.Reals)
ConditionSet(x, Eq(3**(-x)*(2**x - 4**x), 0), S.Reals)
>>> solveset((2**x - 4**x)/(3**x),x,S.Reals)
{0}
```

So who is resonsible for getting the equation into a given form?

Reply:
When the equation is passed via `solveset` , it breaks down to `2**x - 4**x` and `3**(-x)` and the former is easily handled by `transolve`, but equation passed directly in transolve would not yield the result as it does not break the equation.

We can have an `lhs.is_Mul` case that will invoke `_solveset` to handle the equation.

Reply:
> I would raise a NotImplementedError and say something like "This Mul expression should be handled in solveset."

Will raising a `TypeError` or `ValueError` be better because `NotImplementedError` would in a way suggest that the equation can be solved (but this cannot be solved by `transolve` anytime in future).
And yes I would rather add in the docstring that what forms does the `transolve` expect the equation to be.